spec/horoshop_spec.rb

```diff
@@ -3,21 +3,73 @@
 require 'spec_helper'
 
 describe Horoshop do
+  let(:url) { 'http://api.horosop.com' }
+  let(:username) { 'valid_user' }
+  let(:password) { 'valid_password' }
+  let(:invalid_username) { 'invalid_user' }
+  let(:invalid_password) { 'invalid_password' }
+
   describe '#initialize' do
     subject { described_class.new(params) }
 
     context 'when all parameters passed' do
       let(:params) do
-        { url: 'http://api.horosop.com', username: 'user', password: '123456' }
+        { url: url, username: username, password: password }
       end
 
       it { is_expected.to be_a Horoshop }
     end
 
     context 'when parameters' do
-      let(:params) { { url: 'http://api.horosop.com' } }
+      let(:params) { { url: url } }
 
       it { expect { subject }.to raise_error(ArgumentError) }
     end
   end
+
+  describe '#setup_connection and #authorize' do
+    context 'with valid params' do
+      before do
+        stub_request(:post, url)
+          .with(body: { username: username, password: password }.to_json)
+          .to_return(status: 200, body: { status: 'OK', token: 'valid_token' }.to_json)
+      end
+
+      it 'authorize and set a token' do
+        horoshop = Horoshop.new(url: url, username: username, password: password)
+        horoshop.setup_connection
+        expect(horoshop.send(:token)) == ('valid_token')
+      end
+    end
+
+    context 'with invalid params' do
+      before do
+        stub_request(:post, url)
+          .with(body: { username: invalid_username, password: invalid_password })
```

Comment:
Invalid password is enough to the receive error response. Variables **invalid_user** and **invalid_password** is redundant. You can just write

`{ username: username, password: 'invallid' }`